Use HoneySQL to generate INSERT SQL, instead of clojure.java.jdbc

plexus opened the pull request:
Inserts were being done with jdbc/insert-multi!, which internally generates the
necessary SQL statements. This means HoneySQL is bypassed for inserts, and so
HoneySQL features like calling SQL functions with (sql/call ,,,) are unavailable.

This patch instead makes HoneySQL generate the INSERT statements, and executes
them with the lower level jdbc/db-do-prepared-return-keys. When inserting
multiple rows at once it executes separate INSERT statements, rather than a
single statement with multiple VALUES. This behavior is the same with
jdbc/insert-multi!. It's otherwise not possible to retrieve the primary keys of
the inserted rows. (This is a limitation of jdbc/db-do-prepared-return-keys, not
of JDBC itself).

I've added test-cases for db/insert-many! and db/simple-insert-many!, since
those were still missing, as well as an extra test case for db/insert! which
demonstrates the use of an SQL function call.

mazameli commented:
@camsaul: ping in case you missed this

camsaul commented:
Oh yeah. Thanks @mazameli i didn't get an email about this. I'll take a look at this @plexus

plexus commented:
Thanks @camsaul! There are two other open PRs as well, a small one (#12) by @bandresen and a bigger one (#14) by me.

camsaul commented on VERSION-HISTORY.md:
```diff
@@ -1,8 +1,12 @@
-*  Fixed [#6](https://github.com/metabase/toucan/issues/6): Fix hydration for fields that end in `-id`. Credit: [@plexus](https://github.com/plexus)
+* Make `toucan.db/insert!` and friends use HoneySQL, so that you can use `honeysql.core/call` for SQL function calls. ([@plexus](https://github.com/plexus))
+
+### [1.0.3] (May 3rd, 2017)
```
why did you put the brackets around `1.0.3`?

camsaul replied:
nevermind i see that you put the link below

camsaul commented on VERSION-HISTORY.md:
```diff
@@ -1,8 +1,12 @@
+
+### [1.0.3] (May 3rd, 2017)
+
+*  Fixed [#6](https://github.com/metabase/toucan/issues/6): Fix hydration for fields that end in `-id`. ([@plexus](https://github.com/plexus))
```
just wondering what the reason behind changing the format of this was

camsaul commented on test/toucan/db_test.clj:
```diff
@@ -169,9 +170,29 @@
-;; TODO - Test simple-insert-many!
+;; Test simple-insert-many!
+(expect
+ '(4 5)
```
you can just do `[4 5]` here instead of a `'(4 5)` since expectations will consider sequences with the same contents to be the same even if they're vectors instead of lists

camsaul commented on test/toucan/db_test.clj:
```diff
@@ -169,9 +170,29 @@
+   (db/insert-many! User [{:first-name "Grass" :last-name #sql/call [:upper "Hopper"]}
+                          {:first-name "Ko" :last-name "Libri"}])))
+
+(expect
```
what is this supposed to be testing? A quick comment would be useful to explain the purpose of this test

(this goes for other tests as well)

camsaul replied:
e.g. why do we expect this test to fail. It's not clear to me by looking at it why it's supposed to fail

camsaul commented on test/toucan/db_test.clj:
```diff
@@ -169,9 +170,29 @@
+                          {:first-name "Ko" :last-name "Libri"}])))
+
+(expect
+ AssertionError ;; triggered by pre-insert hook
```
Don't feel the need to change this, but for future reference comments on the same lines as code generally are preceded by a single semicolon, e.g.

```clojure
AssertionError ; triggered by pre-insert hook
```

(We try to strictly adhere to the [Clojure Style Guide](https://github.com/bbatsov/clojure-style-guide). The relevant section is [here](https://github.com/bbatsov/clojure-style-guide#one-semicolon-for-margin-comments).)

plexus commented:
Hi @camsaul, thanks for reviewing this. Sorry for the delay, I had some other stuff I had to wrap up first.

I think I addressed your comments. Let me know if there's anything else.

camsaul commented:
Thanks @plexus,
Sorry I haven't got a chance to go over this the last few days but I should have time to take a look at everything tomorrow. 👍
